pkg/mcp/source/source.go

```diff
@@ -158,14 +167,13 @@ type connection struct {
 	queue *internal.UniqueQueue
 }
 
-const DefaultRetryPushDelay = 10 * time.Millisecond
-
 // New creates a new resource source.
 func New(options *Options) *Source {
 	s := &Source{
-		watcher:     options.Watcher,
-		collections: options.CollectionsOptions,
-		reporter:    options.Reporter,
+		watcher:        options.Watcher,
+		collections:    options.CollectionsOptions,
+		reporter:       options.Reporter,
+		requestLimiter: options.RateLimiter,
```

Comment:
This shares a single rate limiter across all connections. It may be difficult to tune fairness. It may be better to have per-connection limits which would require modifying the rate limit interface a bit. Adding an ID parameter is one option which would allow multiplexing multiple connection limits over a single interface. Another option would be to pass a "factory" of sorts as the option rather than a rate limiter instance directly. The caller can then chose to multiplex or create dedicated instances per client.

Reply:
To tackle this added a ratelimiter under internal package which keeps track of Connections and associates them to a rate limiter.